### PR TITLE
[GeoMechanicsApplication] Fix ordering issue for reset displacement

### PR DESCRIPTION
**Pull request by rfaasse**
**📝 Description**
We found an ordering issue with the resetting of the displacements, which was done before executing the processes for Dirichlet conditions, such that prescribed displacements were not overwritten. However, for the reset displacement process, we need the displacements of the elements to not have been reset yet. Therefore, we revert the fix made in https://github.com/KratosMultiphysics/Kratos/pull/11933, but make sure that fixed Dofs are not overwritten to make sure reset displacements and prescribed displacements still work in a single stage.

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 1]
```diff
@@ -15,15 +15,17 @@
...
-#include "custom_utilities/condition_utilities.hpp"
-#include "custom_utilities/element_utilities.hpp"
-#include "custom_utilities/interface_element_utilities.h"
+#include "custom_utilities/node_utilities.h"
 
 namespace Kratos::Python
 {
 
-void AddCustomUtilitiesToPython(pybind11::module&) { namespace py = pybind11; }
+void AddCustomUtilitiesToPython(pybind11::module& m)
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_python/add_custom_utilities_to_python.cpp]: I would suggest to opt for a more descriptive name and to follow the Kratos Style Guide:
```suggestion
void AddCustomUtilitiesToPython(pybind11::module& rModule)
```

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py [hunk 2]
```diff
@@ -57,14 +57,15 @@ def _CalculateTotalDisplacement(self,node, old_total_displacement):
...
         if self._GetSolver().main_model_part.HasNodalSolutionStepVariable(variable):
-            KratosMultiphysics.VariableUtils().SetHistoricalVariableToZero(variable, self._GetSolver().GetComputingModelPart().Nodes)
-            for node in self._GetSolver().GetComputingModelPart().Nodes:
-                new_value = node.GetSolutionStepValue(variable, 0)
-                node.SetSolutionStepValue(variable, 1, new_value)
-
-    def ModifyInitialGeometry(self):
-        # Overrides the base class. Necessary to let reset_displacements function correctly i.c.w. prescribed displacements/rotations.
-        # The reset needs to take place befor the Initialize of the processes, as these will set the Dirichlet condition.
+            zero_vector = Kratos.Array3([0.0, 0.0, 0.0])
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py:60]: I believe you can also write it like this:
```suggestion
            zero_vector = variable.Zero()
```

I'm not sure whether the call operator `()` is needed here.

File: applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.h [hunk 3]
```diff
@@ -77,13 +77,10 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) KratosGeoSettlement
...
 
-        VariableUtils{}.SetHistoricalVariableToZero(rVariable, GetComputationalModelPart().Nodes());
-
-        block_for_each(GetComputationalModelPart().Nodes(),
-                       [&rVariable, SourceIndex, DestinationIndex](auto& node) {
-            node.GetSolutionStepValue(rVariable, DestinationIndex) =
-                node.GetSolutionStepValue(rVariable, SourceIndex);
-        });
+        NodeUtilities::AssignUpdatedVectorVariableToNonFixedComponentsOfNodes(
+            GetComputationalModelPart().Nodes(), rVariable, ZeroVector{3}, 0);
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.h]: Also here, you could opt for asking the zero value from the variable:
```suggestion
            GetComputationalModelPart().Nodes(), rVariable, rVariable.Zero(), 0);
```

File: applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.h [hunk 4]
```diff
@@ -77,13 +77,10 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) KratosGeoSettlement
...
+        NodeUtilities::AssignUpdatedVectorVariableToNonFixedComponentsOfNodes(
+            GetComputationalModelPart().Nodes(), rVariable, ZeroVector{3}, 1);
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.h]: And then the same thing here:
```suggestion
            GetComputationalModelPart().Nodes(), rVariable, rVariable.Zero(), 1);
```

**Review by avdg81 — COMMENTED** [review]
This is a very nice improvement. Thank you so much for catching and fixing it. I only have a few minor suggestions, nothing blocking.

**Review by WPK4FEM — DISMISSED** [review]
Partly formatting.
Partly an doing the variable vector component setting  in C++ i.s.o. python which will be faster.
Partly not reset fixed variable vector components and the order change that is made possible with that.

File: applications/GeoMechanicsApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 1 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_python/add_custom_utilities_to_python.cpp]: Done

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py (hunk 2 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py:60]: Unfortunately it doesn't work (with or without parentheses): `'Kratos.Array1DVariable3' object has no attribute 'Zero'`

File: applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.h (hunk 3 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.h]: Done

File: applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.h (hunk 4 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.h]: Done

File: applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py (hunk 2 above)
Comment (rfaasse) [reply applications/GeoMechanicsApplication/python_scripts/geomechanics_analysis.py:60]: We found that if we export the function to python it does work, but that'd require a core change, which we don't want to do right now

**Review by avdg81 — DISMISSED** [review]
I believe this one is ready to go.

**Review by avdg81 — DISMISSED** [review]
I believe this PR is ready to be merged.

**Review by WPK4FEM — COMMENTED** [review]
Looks o.k. to me

**Review by WPK4FEM — APPROVED** [review]

**Pull request merged**